src/main/java/io/quarkus/gizmo2/impl/BlockCreatorImpl.java

```diff
@@ -1073,17 +1078,18 @@ public void yield(final Expr val) {
     }
 
     public Expr exprHashCode(final Expr expr) {
-        return switch (expr.typeKind()) {
-            case BOOLEAN -> invokeStatic(MethodDesc.of(Boolean.class, "hashCode", int.class, boolean.class), expr);
-            case BYTE -> invokeStatic(MethodDesc.of(Byte.class, "hashCode", int.class, byte.class), expr);
-            case SHORT -> invokeStatic(MethodDesc.of(Short.class, "hashCode", int.class, short.class), expr);
-            case CHAR -> invokeStatic(MethodDesc.of(Character.class, "hashCode", int.class, char.class), expr);
-            case INT -> invokeStatic(MethodDesc.of(Integer.class, "hashCode", int.class, int.class), expr);
-            case LONG -> invokeStatic(MethodDesc.of(Long.class, "hashCode", int.class, long.class), expr);
-            case FLOAT -> invokeStatic(MethodDesc.of(Float.class, "hashCode", int.class, float.class), expr);
-            case DOUBLE -> invokeStatic(MethodDesc.of(Double.class, "hashCode", int.class, double.class), expr);
-            case REFERENCE -> invokeVirtual(MethodDesc.of(Object.class, "hashCode", int.class), expr);
-            case VOID -> Constant.of(0); // null constant
+        String desc = expr.type().descriptorString();
+        return switch (desc.charAt(0)) {
+            case 'Z' -> invokeStatic(MethodDesc.of(Boolean.class, "hashCode", int.class, boolean.class), expr);
+            case 'B' -> invokeStatic(MethodDesc.of(Byte.class, "hashCode", int.class, byte.class), expr);
+            case 'S' -> invokeStatic(MethodDesc.of(Short.class, "hashCode", int.class, short.class), expr);
+            case 'I' -> invokeStatic(MethodDesc.of(Integer.class, "hashCode", int.class, int.class), expr);
+            case 'J' -> invokeStatic(MethodDesc.of(Long.class, "hashCode", int.class, long.class), expr);
+            case 'F' -> invokeStatic(MethodDesc.of(Float.class, "hashCode", int.class, float.class), expr);
+            case 'D' -> invokeStatic(MethodDesc.of(Double.class, "hashCode", int.class, double.class), expr);
+            case 'C' -> invokeStatic(MethodDesc.of(Character.class, "hashCode", int.class, char.class), expr);
+            case 'L', '[' -> invokeStatic(MethodDesc.of(Objects.class, "hashCode", int.class, Object.class), expr);
+            default -> throw impossibleSwitchCase(expr);
```

Comment:
Did you discover some problem with using type kinds that prompted this change?

Reply:
Not really. I originally had code to deal with multidimensionals array here too, so I had to differentiate between objects and arrays. I'll put the code back to its original form.

Reply:
Done.